app/src/main/java/ai/elimu/appstore/synchronization/DownloadApplicationAsyncTask.java

```diff
@@ -190,6 +193,13 @@ protected Integer doInBackground(ApplicationVersion... applicationVersions) {
                         Timber.i(e, "IOException");
                     }
                 }
+
+                /**
+                 * Delete downloaded APK file in case its checksum is invalid
+                 */
+                if (!downloadedApkChecksum.equals(applicationVersion.getChecksumMd5())) {
+                    apkFile.delete();
```

Comment:
Would it be possible to calculate the checksum based on the InputStream instead of the File? If so, could we, instead of deleting corrupt files, prevent them from being stored on the SD card in the first place?

Reply:
Yes it's possible, using checkSumInStream function from this link:
https://stackoverflow.com/a/21710105/3682198
I'll implement this logic.